File: src/algorithms/scan.js
export function calculateSCAN(requests, headStart, diskSize = 200, direction = 'right') {
    let currentHead = headStart;
    let totalSeek = 0;
    const steps = [];

    // Sort requests
    let queue = [...requests].sort((a, b) => a - b);

    // Split into left and right of head
    let left = queue.filter(r => r < headStart).sort((a, b) => b - a); // Descending for left
    let right = queue.filter(r => r >= headStart).sort((a, b) => a - b); // Ascending for right

    let sequence = [];

    if (direction === 'right') {
        sequence = [...right];
        // If we have to change direction, we go to valid end
        if (left.length > 0) {
            sequence.push(diskSize - 1); // Go to end
            sequence = [...sequence, ...left];
        }
    } else {
        sequence = [...left];
        // If we have to change direction
        if (right.length > 0) {
            sequence.push(0); // Go to start
            sequence = [...sequence, ...right];
        }
    }

    // Initial state
    steps.push({
        track: currentHead,
        seek: 0,
        totalSeek: 0,
        requestsLeft: [...requests], // Approximate
        description: "Start"
    });

    let processed = new Set();

    for (let i = 0; i < sequence.length; i++) {
        const target = sequence[i];
        const distance = Math.abs(target - currentHead);

        // Only add distance if we moved
        if (distance > 0) {
            totalSeek += distance;
            currentHead = target;

            // Check if this target was a request
            if (requests.includes(target)) {
                processed.add(target);
            }

            steps.push({
                track: target,
                seek: distance,
                totalSeek: totalSeek,
                requestsLeft: requests.filter(r => !processed.has(r)),
                description: requests.includes(target) ? `Serve ${target}` : `Scan to End/Start`
            });
        }
    }

    return { steps, totalSeek };
}

File: src/algorithms/scan.c
#include <stdlib.h>
#include <stdio.h>

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

int scan(int head, int arr[], int n, int direction, int disk_size) {
    int totalSeek = 0;
    int current = head;

    int sorted[n];
    for(int i = 0; i < n; i++)
        sorted[i] = arr[i];

    qsort(sorted, n, sizeof(int), compare);

    if(direction == 1) { // RIGHT
        for(int i = 0; i < n; i++) {
            if(sorted[i] >= head) {
                totalSeek += abs(current - sorted[i]);
                current = sorted[i];
            }
        }

        totalSeek += abs(current - (disk_size - 1));
        current = disk_size - 1;

        for(int i = n - 1; i >= 0; i--) {
            if(sorted[i] < head) {
                totalSeek += abs(current - sorted[i]);
                current = sorted[i];
            }
        }
    }
    else { // LEFT
        for(int i = n - 1; i >= 0; i--) {
            if(sorted[i] <= head) {
                totalSeek += abs(current - sorted[i]);
                current = sorted[i];
            }
        }

        totalSeek += abs(current - 0);
        current = 0;

        for(int i = 0; i < n; i++) {
            if(sorted[i] > head) {
                totalSeek += abs(current - sorted[i]);
                current = sorted[i];
            }
        }
    }

    return totalSeek;
}